source/extensions/config_subscription/grpc/grpc_mux_failover.h

```diff
@@ -0,0 +1,158 @@
+#pragma once
+
+#include "source/extensions/config_subscription/grpc/grpc_stream.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * This class arbitrates between two config providers of the same GrpcMux -
+ * the primary and the failover. Envoy always prefers fetching config from the
+ * primary source, but if not available, will fetch the config from the failover
+ * source until the primary is again available.
+ *
+ * This class owns the state for the GrpcMux primary and failover streams, and
+ * proxies the gRPC stream functionality to either the primary or the failover config sources.
+ * The failover source is optional and will only be used if passed upon initialization.
+ *
+ * The primary config source is always preferred over the failover source. If the primary
+ * is available, the stream to the failover source will terminated.
+ *
+ * Failover is supported in both
+ * SotW (RequestType = envoy::service::discovery::v3::DiscoveryRequest,
+ *   ResponseType = envoy::service::discovery::v3::DiscoveryResponse), and
+ * Delta-xDS (RequestType = envoy::service::discovery::v3::DeltaDiscoveryRequest,
+ *   ResponseType = envoy::service::discovery::v3::DeltaDiscoveryResponse).
+ * Both the primary and failover streams are either SotW or Delta-xDS.
```

Comment:
One thing coming to my mind, it seems we can do similar failover by the `additional address` of the endpoint https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/endpoint/v3/endpoint_components.proto#envoy-v3-api-msg-config-endpoint-v3-endpoint

Then I begin to think about why we don't reuse it. After I looked the design doc, I understand, you may want some specific failover logic for the xds. (like always reconnect primary xds server per 10 seconds after failover)

Am I understand correctly for the design?

Reply:
Also, thinking about why we didn't improve the aggregate cluster https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/aggregate_cluster#arch-overview-aggregate-cluster for this case

Reply:
This is very different than additional-address or any internal load-balanced hosts (e.g., multiple endpoints in the cluster's list).
The idea is to have a primary (preferred) and a failover options. The preferred option will always be used, but if it is not available the failover will be used. Note that when the primary is available again, Envoy will connect to it and disconnect from the failover.